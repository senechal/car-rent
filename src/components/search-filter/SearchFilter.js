import { useCallback } from 'react';
import { Checkbox, Collapse } from '@reusable';
import { useDispatch, useSelector } from 'react-redux';
import {
  SPECIFICS_LIST,
  SEAT_LIST,
  BAGS_LIST,
} from './constants';
import { applyFilter, removeFilter } from './actions';

const { Group } = Checkbox;

const mapping = (filteredValues) => ({field, value, label}) => {
  const checked = filteredValues.includes(value.toString());
  return (
    <Checkbox value={`${field}:${value}:${label}`} key={label} checked={checked}>{label}</Checkbox>
  )
};

const mapVendors = (filteredValues) => value => {
  const checked = filteredValues.includes(value);
  return (
    <Checkbox value={`vendor:${value}:${value}`} key={value} checked={checked}>{value}</Checkbox>
  )
};

const SearchFilter = () => {
  const { vendors, filters } = useSelector(state => state);
  const dispatch = useDispatch();

  const filtersValues = filters.map(filter => filter.value);

  const handleFilterSelect = useCallback((filterValue) => {
    const [field, value, label] = filterValue.split(':');
    if (filtersValues.includes(value)) {
      dispatch(removeFilter(field, value));
    } else {
      dispatch(applyFilter(field, value, label));
    }
  }, [dispatch, filtersValues]);

  return (
    <Group onChange={handleFilterSelect}>
      <Collapse header="Vendors">
        { vendors.map(mapVendors(filtersValues)) }
      </Collapse>
      <Collapse header="Car Specification">
        { SPECIFICS_LIST.map(mapping(filtersValues)) }
      </Collapse>
      <Collapse header="Seats">
        { SEAT_LIST.map(mapping(filtersValues)) }
      </Collapse>
      <Collapse header="Luggage">
        { BAGS_LIST.map(mapping(filtersValues)) }
      </Collapse>
    </Group>
  )
};

export default SearchFilter;
