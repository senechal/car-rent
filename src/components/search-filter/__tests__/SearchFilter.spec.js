import { useSelector } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import SearchFilter from '../SearchFilter';
import {
  APPLY_FILTER,
  REMOVE_FILTER,
} from '@utils/actionTypes';

const mockVendors = ['VENDOR 1', 'VENDOR 2', 'VENDOR 3'];

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => mockDispatch,
}));

describe('Testing SearchFilter', () => {
  beforeEach(() => {
    useSelector.mockReset();
    mockDispatch.mockReset();
  });
  it('should render correctly', () => {
    useSelector.mockReturnValue({
      vendors: mockVendors,
      filters: [],
    });
    const { container } = render(<SearchFilter />);
    expect(container).toMatchSnapshot();
  });
  it('should render correctly with filters', () => {
    useSelector.mockReturnValue({
      vendors: mockVendors,
      filters: [{value: 'VENDOR 1', field: 'vendor', label: 'VENDOR 1'}],
    });
    const { container } = render(<SearchFilter />);
    expect(container).toMatchSnapshot();
  });
  it('should dispatch applyFilter on empty checkbox change', () => {
    useSelector.mockReturnValue({
      vendors: mockVendors,
      filters: [],
    });
    const filter = {value: 'VENDOR 1', field: 'vendor', label: 'VENDOR 1'};
    const { getAllByTestId } = render(<SearchFilter />);
    const [checkbox] = getAllByTestId('checkbox');
    fireEvent.click(checkbox);
    expect(mockDispatch).toBeCalledWith({
      type: APPLY_FILTER,
      ...filter,
    });
  });
  it('should dispatch removeFilter on checked checkbox change', () => {
    useSelector.mockReturnValue({
      vendors: mockVendors,
      filters: [{value: 'VENDOR 1', field: 'vendor', label: 'VENDOR 1'}],
    });
    const filter = {value: 'VENDOR 1', field: 'vendor'};
    const { getAllByTestId } = render(<SearchFilter />);
    const [checkbox] = getAllByTestId('checkbox');
    fireEvent.click(checkbox);
    expect(mockDispatch).toBeCalledWith({
      type: REMOVE_FILTER,
      ...filter,
    });
  });
});
