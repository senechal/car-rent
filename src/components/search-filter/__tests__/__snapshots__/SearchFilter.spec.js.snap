// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing SearchFilter should render correctly 1`] = `
<div>
  <section>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Vendors
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 1:VENDOR 1"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 1:VENDOR 1"
            />
            VENDOR 1
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 2:VENDOR 2"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 2:VENDOR 2"
            />
            VENDOR 2
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 3:VENDOR 3"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 3:VENDOR 3"
            />
            VENDOR 3
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Car Specification
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="hasAirConditioning:true:With Air-conditioning"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="hasAirConditioning:true:With Air-conditioning"
            />
            With Air-conditioning
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="transmissionType:Automatic:Automatic Transmission"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="transmissionType:Automatic:Automatic Transmission"
            />
            Automatic Transmission
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="transmissionType:Manual:Manual Transmission"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="transmissionType:Manual:Manual Transmission"
            />
            Manual Transmission
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="fuelType:Diesel:Diesel"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="fuelType:Diesel:Diesel"
            />
            Diesel
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Seats
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:2:2 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:2:2 Seats"
            />
            2 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:4:4 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:4:4 Seats"
            />
            4 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:5:5 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:5:5 Seats"
            />
            5 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:5+:5+ Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:5+:5+ Seats"
            />
            5+ Seats
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Luggage
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:2:2 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:2:2 Bags"
            />
            2 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:3:3 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:3:3 Bags"
            />
            3 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:4:4 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:4:4 Bags"
            />
            4 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:5:5 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:5:5 Bags"
            />
            5 Bags
          </label>
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`Testing SearchFilter should render correctly with filters 1`] = `
<div>
  <section>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Vendors
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 1:VENDOR 1"
            data-testid="checkbox"
          >
            <input
              checked=""
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 1:VENDOR 1"
            />
            VENDOR 1
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 2:VENDOR 2"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 2:VENDOR 2"
            />
            VENDOR 2
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="vendor:VENDOR 3:VENDOR 3"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="vendor:VENDOR 3:VENDOR 3"
            />
            VENDOR 3
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Car Specification
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="hasAirConditioning:true:With Air-conditioning"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="hasAirConditioning:true:With Air-conditioning"
            />
            With Air-conditioning
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="transmissionType:Automatic:Automatic Transmission"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="transmissionType:Automatic:Automatic Transmission"
            />
            Automatic Transmission
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="transmissionType:Manual:Manual Transmission"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="transmissionType:Manual:Manual Transmission"
            />
            Manual Transmission
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="fuelType:Diesel:Diesel"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="fuelType:Diesel:Diesel"
            />
            Diesel
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Seats
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:2:2 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:2:2 Seats"
            />
            2 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:4:4 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:4:4 Seats"
            />
            4 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:5:5 Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:5:5 Seats"
            />
            5 Seats
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="seats:5+:5+ Seats"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="seats:5+:5+ Seats"
            />
            5+ Seats
          </label>
        </div>
      </div>
    </div>
    <div>
      <div
        class="rac-collapse-header"
        data-testid="collapse"
      >
        Luggage
        <span
          aria-label="up"
          class="anticon anticon-up rac-collapse-indicator"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="up"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
            />
          </svg>
        </span>
      </div>
      <div
        aria-hidden="false"
        class="rac-collapse-body"
        style="height: auto; overflow: initial;"
      >
        <div
          class="rac-collapse-content"
        >
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:2:2 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:2:2 Bags"
            />
            2 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:3:3 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:3:3 Bags"
            />
            3 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:4:4 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:4:4 Bags"
            />
            4 Bags
          </label>
          <label
            class="rac-checkbox-label"
            data-checkboxvalue="baggageQuantity:5:5 Bags"
            data-testid="checkbox"
          >
            <input
              class="rac-checkbox-input"
              readonly=""
              type="checkbox"
              value="baggageQuantity:5:5 Bags"
            />
            5 Bags
          </label>
        </div>
      </div>
    </div>
  </section>
</div>
`;
